error: No matching export in "App.vue" for import "default" caused by commented // export default {

### Describe the bug

### Version
Vite 2.x, Vue 3.2.x

### Reproduction link
[https://github.com/mariusa/vue-import-error](https://github.com/mariusa/vue-import-error)




### Steps to reproduce
```
git clone git@github.com:mariusa/vue-import-error.git
cd vue-import-error/
npm i
npm run dev
```

### What is expected?
start the app

### What is actually happening?
```
> my-vue-app@0.0.0 dev /home/marius/work/misc/vue-import-error
> vite

 > src/main.js:2:7: error: No matching export in "html:/home/marius/work/misc/vue-import-error/src/App.vue" for import "default"
    2 │ import App from './App.vue'
      ╵        ~~~

error when starting dev server:
Error: Build failed with 1 error:
src/main.js:2:7: error: No matching export in "html:/home/marius/work/misc/vue-import-error/src/App.vue" for import "default"
    at failureErrorWithLog (/home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:1449:15)
    at /home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:1131:28
    at runOnEndCallbacks (/home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:921:63)
    at buildResponseToResult (/home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:1129:7)
    at /home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:1236:14
    at /home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:609:9
    at handleIncomingPacket (/home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:706:9)
    at Socket.readFromStdout (/home/marius/work/misc/vue-import-error/node_modules/esbuild/lib/main.js:576:7)
    at Socket.emit (events.js:315:20)
    at Socket.EventEmitter.emit (domain.js:467:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! my-vue-app@0.0.0 dev: `vite`
npm ERR! Exit status 1
```

---
Removing the commented line
```
// export default {
```
solves this. But why does a comment cause it?

Note:  removing it seems to solve it while still allowing to add it again without crashing, because of node_modules/.vite/ cache -- delete that to reproduce again.


### Reproduction

[https://github.com/mariusa/vue-import-error](https://github.com/mariusa/vue-import-error)

### System Info

```shell
npx: installed 1 in 1.902s

  System:
    OS: Linux 5.13 Fedora 34 (Workstation Edition) 34 (Workstation Edition)
    CPU: (8) x64 AMD Ryzen 7 4700U with Radeon Graphics
    Memory: 3.73 GB / 14.99 GB
    Container: Yes
    Shell: 5.1.0 - /bin/bash
  Binaries:
    Node: 14.16.1 - /usr/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.12 - /usr/bin/npm
  Browsers:
    Chrome: 91.0.4472.114
    Firefox: 88.0.1
```


### Used Package Manager

npm

### Logs

```shell
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'run', 'dev' ]
2 info using npm@6.14.12
3 info using node@v14.16.1
4 verbose run-script [ 'predev', 'dev', 'postdev' ]
5 info lifecycle my-vue-app@0.0.0~predev: my-vue-app@0.0.0
6 info lifecycle my-vue-app@0.0.0~dev: my-vue-app@0.0.0
7 verbose lifecycle my-vue-app@0.0.0~dev: unsafe-perm in lifecycle true
8 verbose lifecycle my-vue-app@0.0.0~dev: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/marius/work/my-vue-app/node_modules/.bin:/usr/local/heroku/bin:/usr/local/heroku/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/marius/.local/bin:/home/marius/bin:/home/marius/.cargo/bin:/home/marius/.local/bin:/home/marius/bin:/home/marius/.cargo/bin
9 verbose lifecycle my-vue-app@0.0.0~dev: CWD: /home/marius/work/my-vue-app
10 silly lifecycle my-vue-app@0.0.0~dev: Args: [ '-c', 'vite' ]
11 silly lifecycle my-vue-app@0.0.0~dev: Returned: code: 1  signal: null
12 info lifecycle my-vue-app@0.0.0~dev: Failed to exec dev script
13 verbose stack Error: my-vue-app@0.0.0 dev: `vite`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:315:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:315:20)
13 verbose stack     at maybeClose (internal/child_process.js:1048:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
14 verbose pkgid my-vue-app@0.0.0
15 verbose cwd /home/marius/work/my-vue-app
16 verbose Linux 5.13.4-200.fc34.x86_64
17 verbose argv "/usr/bin/node" "/usr/bin/npm" "run" "dev"
18 verbose node v14.16.1
19 verbose npm  v6.14.12
20 error code ELIFECYCLE
21 error errno 1
22 error my-vue-app@0.0.0 dev: `vite`
22 error Exit status 1
23 error Failed at the my-vue-app@0.0.0 dev script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/__tests__/scan.spec.ts
+++ b/packages/vite/src/node/__tests__/scan.spec.ts
@@ -3,7 +3,7 @@ import {
   commentRE,
   importsRE,
   multilineCommentsRE,
-  singleCommentsRE
+  singlelineCommentsRE
 } from '../optimizer/scan'
 
 describe('optimizer-scan:script-test', () => {
@@ -116,8 +116,8 @@ describe('optimizer-scan:script-test', () => {
       */`.replace(multilineCommentsRE, '')
     expect(ret).not.toContain('export default')
 
-    singleCommentsRE.lastIndex = 0
-    ret = `//export default { }`.replace(singleCommentsRE, '')
+    singlelineCommentsRE.lastIndex = 0
+    ret = `//export default { }`.replace(singlelineCommentsRE, '')
     expect(ret).not.toContain('export default')
   })
 })

--- a/packages/vite/src/node/__tests__/scan.spec.ts
+++ b/packages/vite/src/node/__tests__/scan.spec.ts
@@ -1,4 +1,10 @@
-import { scriptRE, commentRE, importsRE } from '../optimizer/scan'
+import {
+  scriptRE,
+  commentRE,
+  importsRE,
+  multilineCommentsRE,
+  singleCommentsRE
+} from '../optimizer/scan'
 
 describe('optimizer-scan:script-test', () => {
   const scriptContent = `import { defineComponent } from 'vue'
@@ -101,5 +107,17 @@ describe('optimizer-scan:script-test', () => {
     shouldFailArray.forEach((str) => {
       expect(importsRE.test(str)).toBe(false)
     })
+  })
+
+  test('script comments test', () => {
+    multilineCommentsRE.lastIndex = 0
+    let ret = `/*
+      export default { }
+      */`.replace(multilineCommentsRE, '')
+    expect(ret).not.toContain('export default')
+
+    singleCommentsRE.lastIndex = 0
+    ret = `//export default { }`.replace(singleCommentsRE, '')
+    expect(ret).not.toContain('export default')
   })
 })

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -41,7 +41,7 @@ export const importsRE =
   /(?<!\/\/.*)(?<=^|;|\*\/)\s*import(?!\s+type)(?:[\w*{}\n\r\t, ]+from\s*)?\s*("[^"]+"|'[^']+')\s*(?=$|;|\/\/|\/\*)/gm
 
 export const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
-export const singleCommentsRE = /\/\/.*/g
+export const singlelineCommentsRE = /\/\/.*/g
 
 export async function scanImports(config: ResolvedConfig): Promise<{
   deps: Record<string, string>
@@ -239,7 +239,7 @@ function esbuildScanPlugin(
           // empty singleline & multiline comments to avoid matching comments
           const code = js
             .replace(multilineCommentsRE, '/* */')
-            .replace(singleCommentsRE, '')
+            .replace(singlelineCommentsRE, '')
 
           if (
             loader.startsWith('ts') &&

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -41,6 +41,7 @@ export const importsRE =
   /(?<!\/\/.*)(?<=^|;|\*\/)\s*import(?!\s+type)(?:[\w*{}\n\r\t, ]+from\s*)?\s*("[^"]+"|'[^']+')\s*(?=$|;|\/\/|\/\*)/gm
 
 const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
+const singleCommentsRegex = /\/\/.*/g
 
 export async function scanImports(config: ResolvedConfig): Promise<{
   deps: Record<string, string>
@@ -235,6 +236,10 @@ function esbuildScanPlugin(
               js += content + '\n'
             }
           }
+          // empty singleline & multiline comments to avoid matching comments
+          const code = js
+            .replace(multilineCommentsRE, '/* */')
+            .replace(singleCommentsRegex, '')
 
           if (
             loader.startsWith('ts') &&
@@ -247,8 +252,6 @@ function esbuildScanPlugin(
             // the solution is to add `import 'x'` for every source to force
             // esbuild to keep crawling due to potential side effects.
             let m
-            // empty multiline comments to avoid matching commented out imports
-            const code = js.replace(multilineCommentsRE, '/* */')
             while ((m = importsRE.exec(code)) != null) {
               // This is necessary to avoid infinite loops with zero-width matches
               if (m.index === importsRE.lastIndex) {
@@ -258,11 +261,11 @@ function esbuildScanPlugin(
             }
           }
 
-          if (!js.includes(`export default`)) {
+          if (!code.includes(`export default`)) {
             js += `\nexport default {}`
           }
 
-          if (js.includes('import.meta.glob')) {
+          if (code.includes('import.meta.glob')) {
             return {
               // transformGlob already transforms to js
               loader: 'js',

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -40,8 +40,8 @@ const htmlTypesRE = /\.(html|vue|svelte)$/
 export const importsRE =
   /(?<!\/\/.*)(?<=^|;|\*\/)\s*import(?!\s+type)(?:[\w*{}\n\r\t, ]+from\s*)?\s*("[^"]+"|'[^']+')\s*(?=$|;|\/\/|\/\*)/gm
 
-const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
-const singleCommentsRegex = /\/\/.*/g
+export const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
+export const singleCommentsRE = /\/\/.*/g
 
 export async function scanImports(config: ResolvedConfig): Promise<{
   deps: Record<string, string>
@@ -239,7 +239,7 @@ function esbuildScanPlugin(
           // empty singleline & multiline comments to avoid matching comments
           const code = js
             .replace(multilineCommentsRE, '/* */')
-            .replace(singleCommentsRegex, '')
+            .replace(singleCommentsRE, '')
 
           if (
             loader.startsWith('ts') &&

